app/models/concerns/reviewable_action_builder.rb

```diff
@@ -241,6 +241,70 @@ def perform_convert_to_pm(performed_by, _args)
     create_result(:success, :rejected, [created_by_id], false)
   end
 
+  # Override the Reviewable#perform method to add action logging, and to handle deferred finalization
+  # when using the new reviewable UI refresh.
+  #
+  # @param performed_by [User] The user performing the action.
+  # @param action_id [Symbol] The action being performed.
+  # @param args [Hash] Additional arguments for the action.
+  #
+  # @return [Reviewable::PerformResult] The result of the action.
+  def perform(performed_by, action_id, args = nil)
+    args ||= {}
+    guardian = args[:guardian] || Guardian.new(performed_by)
+
+    # Find which bundle this action belongs to BEFORE executing
+    # (actions may change after execution, e.g., hide_post -> unhide_post)
+    actions = actions_for(guardian, args)
+    bundle_type = nil
+    actions.bundles.each do |bundle|
+      if bundle.actions.any? { |a| a.server_action.to_s == action_id.to_s }
+        # Extract bundle type from bundle ID (e.g., "8311-post-actions" -> "post-actions")
+        bundle_type = bundle.id.split("-", 2).last
+        break
+      end
+    end
+
+    # For old UI actions that aren't in the new separated bundles, use a default bundle
+    bundle_type ||= "legacy-actions"
+
+    # Always skip automatic finalization - we'll handle it ourselves
+    args_for_super = args.dup
+    args_for_super[:skip_finalization] = true
+
+    # Skip the pending notification as well, we'll handle it during finalization
+    self.skip_pending_notification = true
+    result = super(performed_by, action_id, args_for_super)
+    self.skip_pending_notification = false
+
+    if result.success? && result.transition_to
+      reviewable_action_logs.create!(
+        action_key: action_id.to_s,
+        status: result.transition_to,
+        performed_by: performed_by,
+        bundle: bundle_type,
+      )
+    end
+
+    if guardian.can_see_reviewable_ui_refresh? && !SiteSetting.reviewable_old_moderator_actions
```

Comment:
What do you think about moving that block to the original Reviewable#perform model instead of overwriting in this module?

Reply:
Aye, that's worked out reasonably well.

The issue at the moment is that I've had to split up some of the logic in `Reviewable#perform`, notably the `Reviewable.transaction` has now been split up into at least 2 transactions, which is causing the `after_commit` hook to be triggered multiple times. I've experimented with merging the transactions, but it ends up pretty messy to maintain behaviour

(This issue wasn't caused by merging the logic into `Reviewable#perform`, I just removed a workaround for the failing spec that I realised was papering over the deeper problem.)